### ISSUE TITLE
False negative with validateThrows in JavadocMethod

### ISSUE THREAD

**Issue by wltan** [checkstyle/checkstyle#8117]
While designing additional UTs for #7955 I came across some unexpected behavior.

Check documentation: https://checkstyle.org/config_javadoc.html#JavadocMethod

```
D:\checkstyletest>type config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="severity" value="error"/>
    <module name="TreeWalker">
        <module name="JavadocMethod">
            <property name="validateThrows" value="true"/>
        </module>

    </module>
</module>

D:\checkstyletest>type test\*.java 
class Test {

    /**
     * No throws tags.
     *
     * @param x some input
     */
    void foo(int x) {
        if (x > 0) {
            throw new IllegalArgumentException("bar1"); // violation
        }
        if (x < 0) {
            throw new UnsupportedOperationException("bar2"); // violation?
        }
    }

}

D:\checkstyletest>set RUN_LOCALE="-Duser.language=en -Duser.country=US" 

D:\checkstyletest>java -jar "-Duser.language=en -Duser.country=US" checkstyle-8.31-all.jar -c config.xml test\*.java 
Starting audit...
[ERROR] D:\checkstyletest\test\Test.java:10:23: Expected @throws tag for 'IllegalArgumentException'. [JavadocMethod]
Audit done.

```

Expected: There should be 2 violations, one for `IllegalArgumentException` and another for `UnsupportedOperationException`.

This is a different issue from #7473 which is dealing with false positives. This is a false negative.

---

While I haven't figured out the exact conditions necessary to cause this, I suspect that it is due to a bug in this method, as it is simply not returning all the tokens that are necessary to check:
https://github.com/checkstyle/checkstyle/blob/0294924e4066aaa00080a876ab1268a4c8fd7cb3/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java#L728-L755

I'll start working on a fix once this is approved.

**Comment by romani** [comment]
issue is approved, but we need to be careful to not produce false-positives, but regression will show as all.

**Comment by wltan** [comment]
@romani Thanks! Here is the regression for the basic fix https://github.com/wltan/checkstyle/commit/e74993c0dc1cdc21a81f6e50910e8f01b39079c1: https://wltan.github.io/checkstyle-reports/2020-04-18/javadocmethod-traverse/index.html

Will send a proper PR with UTs soon.

**Comment by wltan** [comment]
> we need to be careful to not produce false-positives, but regression will show as all.

@romani There are false positives being added when anonymous class methods are involved e.g.
https://wltan.github.io/checkstyle-reports/2020-04-18/javadocmethod-traverse/apache-ant/index.html#A12
https://wltan.github.io/checkstyle-reports/2020-04-18/javadocmethod-traverse/elasticsearch/index.html#A1

But this is problem is already present in our current release:
```
D:\checkstyletest>type config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="severity" value="error"/>
    <module name="TreeWalker">
        <module name="JavadocMethod">
            <property name="validateThrows" value="true"/>
        </module>

    </module>
</module>

D:\checkstyletest>type test\*.java 
class Test {

    /**
     * No throws tag.
     *
     * @param x some input
     */
    void foo(int x) {
        if (x > 0) {
            throw new IllegalArgumentException("bar1");
        }
        if (x < 0) {
            throw new UnsupportedOperationException("bar2");
        }
    }

    /**
     * No throws tag.
     *
     * @param x some input
     * @return some Runnable
     */
    Runnable foo2(int x) {
        return new Runnable() {
            @Override
            public void run() {
                throw new IllegalArgumentException("");
            }
        };
    }

}

D:\checkstyletest>set RUN_LOCALE="-Duser.language=en -Duser.country=US" 

D:\checkstyletest>java -jar "-Duser.language=en -Duser.country=US" checkstyle-8.31-all.jar -c config.xml test\*.java 
Starting audit...
[ERROR] D:\checkstyletest\test\Test.java:10:23: Expected @throws tag for 'IllegalArgumentException'. [JavadocMethod]
[ERROR] D:\checkstyletest\test\Test.java:27:27: Expected @throws tag for 'IllegalArgumentException'. [JavadocMethod]
Audit done.
```

What this change does is that more of these false positives are being detected because our checking is more thorough this time.

If anything I feel that this problem can be added to the scope of #7473 as we are already dealing with lambdas there.

**Comment by romani** [comment]
fix is merged.
@wltan , thanks a lot for fix.
